geometric/neb.py

```diff
@@ -0,0 +1,2289 @@
+            force_v_p = force_v - np.dot(force_v, tau) * tau
+            if self.climbSet and n in self.climbers:
+                # The climbing image feels no spring forces at all,
+                # and the force in the direction of the tangent is reversed.
+                # Note: We make the choice to change both the plain and the
+                # projected spring force.
+                force_v_p = force_v - 2 * np.dot(force_v, tau) * tau
+                force_s_p *= 0.0
+                force_s *= 0.0
+            # We should be working with gradients
+            grad_s = -1.0 * force_s
+            grad_s_p = -1.0 * force_s_p
+            grad_v = -1.0 * force_v
+            grad_v_p = -1.0 * force_v_p
+            grad_v_p_c[n] = grad_v_p
+            force_s_p_c[n] = force_s
+            force_s_c[n] = force_s
+
+        grad_s_i = self.GlobalIC.calcGrad(xyz, -force_s_c.flatten())
+        grad_v_p_i = self.GlobalIC.calcGrad(xyz, grad_v_p_c.flatten())
+        grad_s_p_i = self.GlobalIC.calcGrad(xyz, -force_s_p_c.flatten())
+        self.set_global_grad(grad_v_i, "potential", "plain")
+        self.set_global_grad(grad_v_p_i, "potential", "projected")
+        self.set_global_grad(grad_s_i, "spring", "plain")
+        self.set_global_grad(grad_s_p_i, "spring", "projected")
+
+    def ComputeGuessHessian(self, blank=False):
+        # self.ComputeSpringHessian()
+        self.spring_hessian_plain = np.zeros((self.nvars, self.nvars), dtype=float)
+        self.spring_hessian_projected = np.zeros(
+            (self.nvars, self.nvars), dtype=float
+        )
+        if not blank:
+            guess_hessian_potential = self.GlobalIC.guess_hessian(
+                self.get_cartesian_all(endpts=True), k=self.params.guessk
+            )
+        else:
+            # guess_hessian_potential *= 0.0
+            guess_hessian_potential = (
+                np.eye(self.spring_hessian_plain.shape[0]) * self.params.guessk
+            )
+        self.guess_hessian_plain = (
+            guess_hessian_potential + self.spring_hessian_plain
+        )
+        self.guess_hessian_projected = (
+            guess_hessian_potential + self.spring_hessian_projected
+        )
+        # Symmetrize
+        self.guess_hessian_plain = 0.5 * (
+            self.guess_hessian_plain + self.guess_hessian_plain.T
+        )
+        self.guess_hessian_projected = 0.5 * (
+            self.guess_hessian_projected + self.guess_hessian_projected.T
+        )
+        self.guess_hessian_plain.flags.writeable = False
+        self.guess_hessian_projected.flags.writeable = False
+        # When plain is set to 1 or 2, we do not project out the perpendicular component of the spring force.
+        if self.plain >= 1:
+            self.guess_hessian_working = self.guess_hessian_plain
+        else:
+            self.guess_hessian_working = self.guess_hessian_projected
+
+    def ComputeMetric(self):
```

Comment:
Could you check if this `ComputeMetric()` function is ever needed by the NEB code?  (What I mean is, are `self.kmats` or `self.metrics` ever used?)  If not, I think this function could be deleted, along with the scipy dependency in setup.py.

Reply:
NEB does not use `self.kmats` and `self.metrics`. I deleted them.